File: lib/layer.js
var Npc = require("./npc.js"),
    Playable = require("./playable.js"),
    Actionable = require("./actionable.js");

function Layer(layer_options, map){
  this.name = layer_options.name;
  this.type = layer_options.type;
  this.properties = layer_options.properties || {};

  this.opacity = layer_options.opacity;
  this.visible = layer_options.visible;

  this.data = layer_options.data;
  this.objects = {};

  this.x = layer_options.x;
  this.y = layer_options.y;
  this.width = layer_options.width;
  this.height = layer_options.height;

  this.map = map;

  if(this.is_objectgroup()){
    this._initiate_objects(layer_options.objects);
  }
}

Layer.prototype.is_tilelayer = function(){
  return this.type == "tilelayer";
};

Layer.prototype.is_objectgroup = function(){
  return this.type == "objectgroup";
};

Layer.prototype._initiate_objects = function(objects){
  var _this = this;
  for (var i = objects.length - 1; i >= 0; i--) {
    var object = objects[i];
    if(object.type.toLowerCase() == "player"){
      _this.map.player_prototype = object;
      var player = new Playable(object, _this.map);
    }else if(object.type.toLowerCase() == "npc"){
      var npc = new Npc(object, _this.map);
      _this.map.npcs[npc.name] = npc;
    }else if(object.type.toLowerCase() == "actionable"){
      var actionable = new Actionable(object, _this.map);
      _this.map.actionable[actionable.name] = actionable;
    }
  };
};


module.exports = Layer;

File: json_map/layer.go
package json_map

type Layer struct {
  Data    []int     `json:"data"`
  Height  float32   `json:"height"`
  Width   float32   `json:"width"`
  Name    string    `json:"name"`
  Opacity float32   `json:"opacity"`
  Type    string    `json:"type"`
  Visible bool      `json:"visible"`
  X       float32   `json:"x"`
  Y       float32   `json:"y"`
  Sprites []*Sprite `json:"objects"`
}

func (l *Layer) IsTileLayer() bool {
  return l.Type == "tilelayer"
}

func (l *Layer) IsObjectGroup() bool {
  return l.Type == "objectgroup"
}
